### PR TITLE
ENH: Outsource ``apply()`` from transform objects

### LINKED ISSUE #192: Offsource `.apply()` as an independent operation
(low priority)

### PR DESCRIPTION
**Pull request by jmarabotto**
I have joined CHUV (Lausanne, Switzerland) for a 5 month placement and am working with Oscar Esteban on this project. We worked through this problem together as an example of how I can contribute and resolve issues, as I am somewhat new to GitHub. Eventually, this Resolves #192 - but for now, this acts as an example and further changes are needed :)

### CODE REVIEW

**Comment by effigies** [comment]
I assume this is intended to close https://github.com/nipy/nitransforms/issues/192? If so, please add "Resolves #192" or similar to your top post.

**Comment by oesteban** [comment]
Okay, with #197 out of the way we can focus on finalizing this. @effigies -- I'm working with Julien on this one, I will walk him through the process of preparing the PR we will let you know when this is ready for you to eyeball it.

File: nitransforms/linear.py [hunk 1]
```diff
@@ -402,119 +402,6 @@ def to_filename(self, filename, fmt="X5", moving=None):
...
-            (xcoords.shape[0], len(self)), dtype=output_dtype, order="F"
-        )
-
-        dataobj = (
-            np.asanyarray(spatialimage.dataobj, dtype=input_dtype)
-            if spatialimage.ndim in (2, 3)
-            else None
-        )
-
-        for t, xfm_t in enumerate(self):
```
Comment (oesteban) [inline nitransforms/linear.py:488]: I've been holding on this one - the new implementation is 4D, while the old implementation was 3D+t (meaning that you could not resample off-gried on the fourth dimension).

We should make sure we meet today @effigies, @jmarabotto and discuss this.

cc/ @sgiavasis who may be interested in learning more.

File: nitransforms/nonlinear.py [hunk 2]
```diff
@@ -29,6 +30,11 @@ class DenseFieldTransform(TransformBase):
 
     __slots__ = ("_field", "_deltas")
 
+    @property
+    def ndim(self):
+        """Access the dimensions of this Desne Field Transform."""
```
Comment (oesteban) [inline nitransforms/nonlinear.py:35]:
```suggestion
        """Get the dimensions of the transform."""
```

File: nitransforms/nonlinear.py [hunk 3]
```diff
@@ -239,6 +244,12 @@ class BSplineFieldTransform(TransformBase):
 
     __slots__ = ['_coeffs', '_knots', '_weights', '_order', '_moving']
 
+    @property
+    def ndim(self):
+        """Access the dimensions of this BSpline."""
```
Comment (oesteban) [inline nitransforms/nonlinear.py:249]:
```suggestion
        """Get the dimensions of the transform."""
```

File: nitransforms/resampling.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    data = np.asanyarray(spatialimage.dataobj)
+
+    if data.ndim == 4 and data.shape[-1] != len(transform):
+        raise ValueError("The fourth dimension of the data does not match the tranform's shape.")
+
+    if data.ndim < transform.ndim:
+        data = data[..., np.newaxis]
+    
+    if hasattr(transform, 'to_field') and callable(transform.to_field):
```
Comment (oesteban) [inline nitransforms/resampling.py:97]:
```suggestion
    # For model-based nonlinear transforms, generate the corresponding dense field
    if hasattr(transform, 'to_field') and callable(transform.to_field):
```

**Review by oesteban — COMMENTED** [review]
Let's finalize this today.

File: nitransforms/resampling.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    if data.ndim == 4 and data.shape[-1] != len(transform):
+        raise ValueError(
+            "The fourth dimension of the data does not match the tranform's shape."
+        )
+
+    if data.ndim < transform.ndim:
+        data = data[..., np.newaxis]
+
+    if hasattr(transform, "to_field") and callable(transform.to_field):
```
Comment (oesteban) [inline nitransforms/resampling.py:101]:
```suggestion
    # For model-based nonlinear transforms, generate the corresponding dense field
    if hasattr(transform, "to_field") and callable(transform.to_field):
```

File: nitransforms/nonlinear.py [hunk 6]
```diff
@@ -167,19 +163,22 @@ def map(self, x, inverse=False):
...
-            map_coordinates(
-                self._field[..., i],
-                ijk.T,
-                order=3,
-                mode="constant",
-                cval=np.nan,
-                prefilter=True,
-            ) for i in range(self.reference.ndim)
-        )).T
+        return np.vstack(
```
Comment (effigies) [inline nitransforms/nonlinear.py:169]:
```suggestion
        new_map = np.vstack(
```

**Pull request merged**